src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -496,14 +513,113 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
+        // take/return AbstractFloats, and AbstractFloats cannot be stored in
+        // variables, so the code cannot just inject a simple utility function
+        // at the top of the shader, instead this snippet needs to be inlined
+        // everywhere the test needs to return an AbstractFloat.
+        //
+        // select is used below, since ifs are not available during  constant
+        // eval. This has the side effect of short-circuiting doesn't occur, so
+        // both sides of the select have to evaluate and be valid. This is why
+        // handling subnormals is difficult implement.
+        //
+        // This snippet implements FTZ for subnormals to bypass the need for
+        // complex subnormal specific logic.
+        //
+        // Expressions resulting in subnormals
+        // can still be reasonably tested, since this snippet will return 0 with
+        // the correct sign, which is always in the acceptance interval for a
+        // subnormal result, since an implementation may FTZ.
+        //
+        // Document for the snippet is included here in this code block, since
+        // shader length affects compilation time  significantly on some
+        // backends.
+        //
+        // Snippet with documentation:
+        //   const kExponentBias = 1022;
+        //
+        //   // Detect if the value is zero or subnormal, so that FTZ behaviour
+        //   // can occur
+        //   const subnormal_or_zero : bool = (${expr} <= ${kValue.f64.subnormal.positive.max}) && (${expr} >= ${kValue.f64.subnormal.negative.min});
+        //
+        //   // MSB of the high_u32 is 1 if the value is negative, otherwise 0
+        //   // Extract the sign bit early, so that abs() can be used with
+        //   // frexp() to simplify a bunch of logic later
+        //   const sign_bit : u32 = select(0, 0x80000000, ${expr} < 0);
+        //
+        //   // Use frexp() to obtain the exponent and fractional parts, and
+        //   // then perform FTZ if needed
+        //   const f = frexp(abs(${expr}));
+        //   const f_fract = select(f.fract, 0, subnormal_or_zero);
+        //   const f_exp = select(f.exp, -kExponentBias, subnormal_or_zero);
+        //
+        //   // The exponent is stored in bits [20..31] of high_u32, special
+        //   // case for zero.
+        //   const exponent_bits : u32 = (f_exp + kExponentBias) << 20;
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, masking off
+        //   // unneeded bits.
+        //   const high_mantissa_bits : u32 = (u32(f_fract * (1 << 20))) & 0x000fffff;
+        //
+        //   // Subtract out the high bits from the mantissa.
+        //   // const low_mantissa = f_fract - floor(f_fract * (1<< 21)) / (1 << 21);
+        //
+        //   // Extract the low 32 bits of the mantissa as an u32
+        //   // The calculated lower mantissa may or may not have a leading 1 in
```

Comment:
I can't figure out why that leading 1 should be there, except by a numerical error in the calculation somewhere along the way.

The division `  / (1 << 21)` looks fishy to me. See below.

Reply:
I agree